File: client/src/model/Pagination.ts
import { UrlPagination } from "./Url";


export type Pagination = {
    total: number;
    limit: number;
    offset: number;
    page: number;
    pages: number;
}

export function emptyPagination(): Pagination {
    return {
        total: 0,
        limit: 64,
        offset: 0,
        page: 1,
        pages: 0
    }
}


export function extractNextPagination(pagination: UrlPagination): Pagination {
    return {
        total: pagination.total + 1,
        limit: pagination.limit,
        offset: pagination.offset,
        page: pagination.page,
        pages: pagination.pages
    }
}

File: server/src/schemas/pagination.py
from pydantic import BaseModel, model_validator
from typing import Generic, TypeVar, List, Optional
T = TypeVar("T", bound=BaseModel)


class Pagination(BaseModel, Generic[T]):

    total: int
    limit: int
    offset: int
    page: Optional[int] = None
    pages: Optional[int] = None
    results: List[T]

    @model_validator(mode="after")
    def compute_pages(self):
        self.page = (self.offset // self.limit) + 1
        self.pages = (self.total + self.limit - 1) // self.limit
        return self